### PR TITLE
User Registration and login functionality

### PR DESCRIPTION
**Pull request by MarkB-Mugoh**
Implementation of the user registration and login functionalities and responses sent back to the user as JSON objects

### CODE REVIEW

**Review by abdiomari — APPROVED** [review]

File: src/main/java/com/seed/api/core/UserService.java [hunk 1]
```diff
@@ -1,29 +1,136 @@
...
+        this.setServicesRoutes(router);
     }
 
+    /**
+     * Handler for user registration.
+     * @param routingContext The routing context.
+     */
+    private void handleUserRegistration(final RoutingContext routingContext) {
+        // Extract user registration data from the request body
+        JsonObject registrationData = routingContext.getBodyAsJson();
```
Comment (otkothtuk) [inline src/main/java/com/seed/api/core/UserService.java:65]: routingContext.body().asJsonObject();

File: src/main/java/com/seed/api/core/UserService.java [hunk 2]
```diff
@@ -1,29 +1,136 @@
...
+
+        // Extract individual fields from the registration data
+        String fullName = registrationData.getString("full_name");
+        String phoneNumber = registrationData.getString("phone_number");
+        String email = registrationData.getString("email");
+        String password = registrationData.getString("password");
+        String confirmPassword = registrationData.getString("confirm_password");
```
Comment (otkothtuk) [inline src/main/java/com/seed/api/core/UserService.java:72]: You don't need confirm password. Its the front end that needs it for validation.

File: src/main/java/com/seed/api/core/UserService.java [hunk 3]
```diff
@@ -1,29 +1,136 @@
...
+                .put("phone_number", phoneNumber)
+                .put("email", email);
+
+            routingContext.response()
+                .setStatusCode(SUCCESS_CODE)
+                .putHeader("Content-Type", "application/json")
+                .end(response.encode());
+        }
+    }
+
```
Comment (otkothtuk) [inline src/main/java/com/seed/api/core/UserService.java:110]: protected boolean isValid(final String val) {
   return val != null && !val.trim().isEmpty();
}

File: src/main/java/com/seed/api/core/UserService.java [hunk 4]
```diff
@@ -1,29 +1,136 @@
...
+        // Extract individual fields from the registration data
+        String fullName = registrationData.getString("full_name");
+        String phoneNumber = registrationData.getString("phone_number");
+        String email = registrationData.getString("email");
+        String password = registrationData.getString("password");
+        String confirmPassword = registrationData.getString("confirm_password");
+
+        // Validate the registration data
+        //(e.g., check for required fields, email format, password match, etc.)
+        if (fullName == null || fullName.isEmpty()
```
Comment (otkothtuk) [inline src/main/java/com/seed/api/core/UserService.java:76]: this.isValid(fullName) && this.isValid(phoneNumber)

File: src/main/java/com/seed/api/core/UserService.java [hunk 5]
```diff
@@ -1,29 +1,136 @@
...
+        || email == null || email.isEmpty()
+        || password == null || password.isEmpty()
+        || confirmPassword == null || confirmPassword.isEmpty()) {
+            // If any of the required fields are missing, return an error
+            JsonObject errorResponse = new JsonObject()
+                .put("message", "Please fill in all the required fields.");
+            routingContext.response()
+                .setStatusCode(ERR_BAD_REQUEST)
+                .putHeader("Content-Type", "application/json")
+                .end(errorResponse.encode());
```
Comment (otkothtuk) [inline src/main/java/com/seed/api/core/UserService.java:87]: .end(this.getResponse(ERR_BAD_REQUEST, errorResponse).encode());

File: src/main/java/com/seed/api/core/UserService.java [hunk 6]
```diff
@@ -1,29 +1,136 @@
...
+        } else if (!password.equals(confirmPassword)) {
```
Comment (otkothtuk) [inline src/main/java/com/seed/api/core/UserService.java:88]: This should be a front end work.

**Review by otkothtuk — APPROVED** [review]

**Pull request merged**